### DISCUSSION TITLE
Compile and bundle a requested file in a Webpack Plugin

### DISCUSSION THREAD

**Discussion by LukasBombach** [webpack/webpack#18078] [Q&A]
I am trying to write a webpack plugin that can compile a specific file and return its compiled source instead of the requested module. Sounds weird? What I mean is this:

`magic_module.ts`
```ts
function log() {
  console.log("hello world");
}
```

should be transformed to 

```js
module.exports = "function log(){ console.log(\"hello world\"); }"
```

when loaded.

The Background is that I want to load files and render their contens as inline js in a webpage, so

```tsx
import React from "react";

const myModuleSource = await import ("./magic_module");

const InlineScriptComponent = () => {
  return <script dangrouslySetInnerHtml={{ __html: myModuleSource }} />
}
```

I'm thinking I can write a webpack plugin, that would tap into the normal module resolution and compile a file and return the compiled results instead of the module. 

```js
class CompileToStringPlugin {
  constructor() {
    this.PLUGIN_NAME = this.constructor.name;
  }

  apply(compiler) {
    if (!this.options.isServer) {
      compiler.hooks.normalModuleFactory.tap(this.PLUGIN_NAME, nmf => {

        // what is the actual hook that I need to tap into?
        nmf.hooks.beforeResolve.tap(this.PLUGIN_NAME, result => {

          // how can I use webpack here to compile the file result.request
          // with its dependencies to and get the resulting source
          // so that I can return the sourcecode as export
          const compiledSourceCode = compileAndBundle(result.request);

          return `module.exports = ${JSON.stringify(compiledSourceCode)}`;
        });
      });
    }
  }
}
```

But I have two questions:

1. I am pretty sure that `beforeResolve` is not the correct hook here, because I cannot return the contents of a module here, but I am not sure which hook I really need to tap into 
2. I cannot figure out how to actually compile the file here. I can find all sorts of methods that will give me the original source of the file, but I need to actually compile the file _and its dependecies_ to a sort of small bundle and then get the resulting source

**Comment by alexander-akait** [comment]
Hello, why do not use loader here?
